tests/test_update_ssh_keys.rs

```diff
@@ -0,0 +1,622 @@
+    fn del_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str) {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        for key in aks.remove_keys(testname) {
+            if let AuthorizedKeyEntry::Valid {
+                ref key,
+            } = key
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                // must write before checking if the file exists, because before writing,
+                // authorized_keys.d is not in sync.
+                aks.write().expect("failed to update authorized keys directory");
+
+                let authkeyone: PathBuf = PathBuf::from(format!("{}/authorized_keys.d/{}", ssh_dir, testname));
+
+                assert!(!authkeyone.exists());
+
+                let mut svec: Vec<&str> = Vec::new();
+                svec.push(pubkeyname);
+                assert_has_keys(ssh_dir, svec, true);
+            }
+        }
+
+        // In contrast to write, sync can run only once after the loop.
+        aks.sync().expect("failed to update authorized keys");
+    }
+
+    fn setup_tests() -> TestContext {
+        let ssh_dir_path: PathBuf = TempDir::new(SSH_DIR_PREFIX).unwrap().into_path();
+
+        for (name, text) in &*TEST_KEYS {
+            let pub_path = ssh_dir_path.join(format!("{}.pub", name));
+
+            let mut pubfile = File::create(pub_path.clone()).expect(format!("unable to create a file {}", pub_path.to_str().unwrap_or_default()).as_str());
+            let _ = pubfile.write_all(text.as_bytes());
+        }
+
+        test_no_keys(ssh_dir_path.to_str().unwrap());
+        test_first_run(ssh_dir_path.to_str().unwrap());
+
+        TestContext {
+            ssh_dir: ssh_dir_path,
+        }
+    }
+
+    fn test_no_keys(ssh_dir: &str) {
+        let aks = open_authorized_keys(ssh_dir);
+
+        aks.write().expect("failed to update authorized keys directory");
+        assert!(format!("{}", aks.sync().unwrap_err().kind()).contains("no keys found"));
+
+        let authkeysdir: PathBuf = PathBuf::from(format!("{}/authorized_keys.d", ssh_dir));
+        assert!(authkeysdir.is_dir());
+
+        let authkeys: PathBuf = PathBuf::from(format!("{}/authorized_keys", ssh_dir));
+        assert!(!authkeys.exists());
+    }
+
+    fn test_first_run(ssh_dir: &str) {
+        // It is necessary to clean up before running the first run test,
+        // because old_authorized_keys will be only created when authorized_keys.d
+        // does not exist.
+        _ = fs::remove_dir_all(format!("{}/authorized_keys.d", ssh_dir));
```

Comment:
```suggestion
```
With the suggested fixes there shouldn't be any old state and the test can do the same as the python test does.